configtype/time.go

```diff
@@ -0,0 +1,164 @@
+package configtype
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/invopop/jsonschema"
+)
+
+type timeType int
+
+const (
+	timeTypeZero timeType = iota
+	timeTypeFixed
+	timeTypeRelative
+)
+
+// Time is a wrapper around time.Time that should be used in config
+// when a time type is required. We wrap the time.Time type so that
+// the spec can be extended in the future to support other types of times
+type Time struct {
+	typ      timeType
+	time     time.Time
+	duration Duration
+}
+
+func NewTime(t time.Time) Time {
+	return Time{
+		typ:  timeTypeFixed,
+		time: t,
+	}
+}
+
+func ParseTime(s string) (Time, error) {
+	var t Time
+	var err error
+	switch {
+	case timeNowRegexp.MatchString(s):
+		t.typ = timeTypeRelative
+		t.duration = NewDuration(0)
+	case timeRFC3339Regexp.MatchString(s):
+		t.time, err = time.Parse(time.RFC3339, s)
+		if t.time.IsZero() {
+			t.typ = timeTypeZero
+		} else {
+			t.typ = timeTypeFixed
+		}
+	case dateRegexp.MatchString(s):
+		t.typ = timeTypeFixed
+		t.time, err = time.Parse(time.DateOnly, s)
+	case baseDurationRegexp.MatchString(s), humanRelativeDurationRegexp.MatchString(s):
+		t.typ = timeTypeRelative
+		t.duration, err = ParseDuration(s)
+	default:
+		return t, fmt.Errorf("invalid time format: %s", s)
+	}
+
+	return t, err
+}
+
+var (
+	timeRFC3339Pattern = `^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(.(\d{1,9}))?(Z|((-|\+)\d{2}:\d{2}))$`
+	timeRFC3339Regexp  = regexp.MustCompile(timeRFC3339Pattern)
+
+	timeNowPattern = `^now$`
+	timeNowRegexp  = regexp.MustCompile(timeNowPattern)
+
+	datePattern = `^\d{4}-\d{2}-\d{2}$`
+	dateRegexp  = regexp.MustCompile(datePattern)
+
+	timePattern = patternCases(
+		timeNowPattern,
+		timeRFC3339Pattern,
+		datePattern,
+		baseDurationPattern,
+		humanRelativeDurationPattern,
+	)
+)
+
+func (Time) JSONSchema() *jsonschema.Schema {
+	return &jsonschema.Schema{
+		Type:    "string",
+		Pattern: timePattern,
+		Title:   "CloudQuery configtype.Time",
+	}
+}
+
+func (t *Time) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	tim, err := ParseTime(s)
+	if err != nil {
+		return err
+	}
+
+	*t = tim
+	return nil
+}
+
+func (t Time) MarshalJSON() ([]byte, error) {
+	switch t.typ {
```

Comment:
I think we can keep the original string that we got in `ParseTime` and return that instead, WDYT?

Reply:
[263750b](https://github.com/cloudquery/plugin-sdk/pull/1905/commits/263750b142816f645bb9b548586602394e87cbc2), but unclear what the handling for an uninitialized Time or Duration should be. Currently it just marshals to `""`

Reply:
Also, as a side-effect Time is no longer == comparable in a useful way

Reply:
> [263750b](https://github.com/cloudquery/plugin-sdk/pull/1905/commits/263750b142816f645bb9b548586602394e87cbc2), but unclear what the handling for an uninitialized Time or Duration should be. Currently it just marshals to `""`

I think we need to decide that and pick whatever we want to happen, we can

1. Error out on uninitialized
2. Return `MarshalJSON` of the zero value of `time.Time`
3. Return `MarshalJSON` of the zero value of `time.Duration`
4. Return `null`
5. Maybe more options

Not sure what's the best option but it needs to be consistent with `UnmarshalJSON`.
We can see what happens for `time.Time` and `time.Duration` zero values

Reply:
> Also, as a side-effect Time and Duration are no longer == comparable in a useful way

How could you compare Time and Duration before? Where do you need this comparison and how will it be used?

Reply:
Duration was comparable with == (as it was just a struct) containing a time.Duration. I'm unsure if that was used anywhere but removing that is a breaking change. Adding days,months,etc to the struct maintained comparability for equal values but adding the `input` string field breaks that for anything besides values that had the same input string even if they're otherwise equivalent (including whitespace differences).

I guess Time wasn't usefully comparable with `==` as while time.Time is `==` comparable doing doesn't handle timezones and so is not recommended.

As for where it is needed, it isn't exactly for time, but the duration change is breaking change in the API (unclear if anything actually uses it though)

Reply:
Ok so:
1. I don't think we should modify the existing duration config type. Relaxing the parsing is also a breaking change
2. Even if we did modify the existing type, we can used the parsed value of the duration for `Equals`

Reply:
Regarding 2., that still will have broken `==` comparison, although again unclear if that is used externally or if that even counts as an explicit part of the API to be broken

Reply:
Regarding default behavior, returning the zero value of time seems sane as it'll allow the current IsZero check to work. Although that is `"0001-01-01T00:00:00Z"`

https://go.dev/play/p/qI6SprLcw5p